server.go

```diff
@@ -597,12 +640,20 @@ func (srv *Server) serveTCPConnection(ctx *shutdownContext, conn *coapNet.Conn)
 
 // serveListener starts a TCP listener for the server.
 func (srv *Server) serveTCPListener(l Listener) error {
+	srv.doneLock.Lock()
+	if srv.doneChan != nil {
+		return fmt.Errorf("server already serve connections")
```

Comment:
Please could you always use `defer` for unlocking? This path does not unlock. Consider using [Event](https://godoc.org/google.golang.org/grpc/internal/grpcsync).